### PR TITLE
Huggingface optimum prepackaged server

### LINKED ISSUE #4082: Create huggingface optimum transformer prepackaged server
As part of the Huggingface Optimum Transformer collaboration this encompasses creating a pre-packaged server for huggingface models that integrate with the hugginface hub.

### PR DESCRIPTION
**Pull request by axsaucedo**
**Which issue(s) this PR fixes**:

Fixes #4082

Introduces MLServer Runtime: https://github.com/SeldonIO/MLServer/pull/573

# HuggingFace Server

Thanks to our collaboration with the HuggingFace team you can now easily deploy your models from the [HuggingFace Hub](https://huggingface.co/models) with Seldon Core.

We also support the high performance optimizations provided by the [Transformer Optimum framework](https://huggingface.co/docs/optimum/index).

## Pipeline parameters

The parameters available:

| Name | Description |
| ---- | ----------- |
| `task` | The transformer pipeline task |
| `pretrained_model` | The name of the pretrained model in the Hub |
| `pretrained_tokenizer` | Transformer name in Hub if different to the one provided with model |
| `optimum_model` | Boolean to enable loading model with Optimum framework |

## Simple Example

You can deploy a HuggingFace model by providing parameters to your [pipeline](https://huggingface.co/docs/transformers/main_classes/pipelines).

```yaml
apiVersion: machinelearning.seldon.io/v1alpha2
kind: SeldonDeployment
metadata:
  name: gpt2-model
spec:
  protocol: v2
  predictors:
  - graph:
      name: transformer
      implementation: HUGGINGFACE_SERVER
      parameters:
      - name: task
        type: STRING
        value: text-generation
      - name: pretrained_model
        type: STRING
        value: distilgpt2
    name: default
    replicas: 1
```

## Quantized & Optimized Models with Optimum

You can deploy a HuggingFace model loaded using the Optimum library by using the `optimum_model` parameter

```yaml
apiVersion: machinelearning.seldon.io/v1alpha2
kind: SeldonDeployment
metadata:
  name: gpt2-model
spec:
  protocol: v2
  predictors:
  - graph:
      name: transformer
      implementation: HUGGINGFACE_SERVER
      parameters:
      - name: task
        type: STRING
        value: text-generation
      - name: pretrained_model
        type: STRING
        value: distilgpt2
      - name: optimum_model
        type: BOOL
        value: true
    name: default
    replicas: 1
```

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **axsaucedo**
You can assign the PR to them by writing `/assign @axsaucedo` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
@axsaucedo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 027bee0d220c861b4d1081478ef9bd574243e3ed | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4081-integration-kbz8b) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: operator/controllers/seldondeployment_prepackaged_servers.go
Comment (adriangonz) [inline operator/controllers/seldondeployment_prepackaged_servers.go:241]: Would it be simpler to add this one to this file instead: https://github.com/SeldonIO/seldon-core/blob/master/operator/controllers/mlserver.go? 

It should be just a matter of adding the Huggingface server one to this `switch`:

https://github.com/SeldonIO/seldon-core/blob/41738bf547eef2a50a34bd24f5feadd957debbea/operator/controllers/mlserver.go#L234-L250

And then also to this `if` condition:

https://github.com/SeldonIO/seldon-core/blob/41738bf547eef2a50a34bd24f5feadd957debbea/operator/controllers/seldondeployment_prepackaged_servers.go#L423

**Comment by seldondev** [comment]
@axsaucedo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 6f0109a64375b92d48380926985831a843d0933c | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4081-integration-gp9gj) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: operator/controllers/seldondeployment_prepackaged_servers.go
Comment (axsaucedo) [reply operator/controllers/seldondeployment_prepackaged_servers.go:241]: Good shout, yes I agree it would fit better on the mlserver.go file, and adding it to this getmlserverenvvars function

**Comment by seldondev** [comment]
@axsaucedo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 6f0109a64375b92d48380926985831a843d0933c | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4081-integration-kxv87) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: operator/controllers/seldondeployment_prepackaged_servers.go
Comment (axsaucedo) [reply operator/controllers/seldondeployment_prepackaged_servers.go]: This now has been updated - looks much better, let me know if any further thoughts

**Review by adriangonz — APPROVED** [review]
This looks great! I think it should be ready to go @axsaucedo :+1:

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[adriangonz](https://github.com/SeldonIO/seldon-core/pull/4081#pullrequestreview-968892579 "Approved")*

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [adriangonz]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
